src/Internal/Plutip/Spawn.purs

```diff
@@ -38,53 +45,67 @@ data NewOutputAction = NoOp | Success | Cancel
 -- | `spawn`, but with ability to wait for program startup, using a callback
 -- | returning a `NewOutputAction`, or to kill the program depending on its
 -- | output.
-spawnAndWaitForOutput
+spawn
   :: String
   -> Array String
   -> SpawnOptions
-  -> (String -> NewOutputAction)
-  -> Aff ChildProcess
-spawnAndWaitForOutput cmd args opts filter =
-  makeAff (spawnAndWaitForOutput' cmd args opts filter)
+  -> Maybe (String -> NewOutputAction)
+  -> Aff ManagedProcess
+spawn cmd args opts mbFilter =
+  makeAff (spawn' cmd args opts mbFilter)
 
-spawnAndWaitForOutput'
+spawn'
   :: String
   -> Array String
   -> SpawnOptions
-  -> (String -> NewOutputAction)
-  -> (Either Error ChildProcess -> Effect Unit)
+  -> Maybe (String -> NewOutputAction)
+  -> (Either Error ManagedProcess -> Effect Unit)
   -> Effect Canceler
-spawnAndWaitForOutput' cmd args opts filter cont = do
-  child <- spawn cmd args opts
-  ref <- Ref.new (Just "")
-  ChildProcess.onExit child $ const do
-    output <- Ref.read ref
+spawn' cmd args opts mbFilter cont = do
+  child <- ChildProcess.spawn cmd args opts
+  closedAVar <- AVar.empty
+  let mp = ManagedProcess child closedAVar
+  interface <- RL.createInterface (stdout child) mempty
+  outputRef <- Ref.new ""
+  ChildProcess.onClose child \code -> do
+    RL.close interface
+    void $ AVar.tryPut code closedAVar
+    output <- Ref.read outputRef
     cont $ Left $ error $
-      "Process " <> cmd <> " exited. Output:\n" <> fold output
-  onDataString (stdout child) Encoding.UTF8
-    \str -> do
-      output <- Ref.modify (map (_ <> str)) ref
-      case filter <$> output of
-        Just Success -> do
-          -- Set to Nothing to prevent future updates
-          Ref.write Nothing ref
-          cont (pure child)
-        Just Cancel -> do
-          Ref.write Nothing ref
-          kill SIGINT child
-          cont $ Left $ error
-            $ "Process cancelled because output received: " <> str
-        _ -> pure unit
+      "Process " <> cmd <> " exited. Output:\n" <> output
+
+  -- Ideally we close the interface instead of detaching the listener
+  -- but it seems to issue a close on the output stream of the process too
+  -- which breaks some processes
+  case mbFilter of
+    Nothing -> cont (pure mp)
+    Just filter -> do
+      flip RL.setLineHandler interface
+        \str -> do
+          output <- Ref.modify (_ <> str <> "\n") outputRef
+          case filter output of
+            Success -> do
+              clearLineHandler interface
+              cont (pure mp)
+            Cancel -> do
+              kill SIGINT child
+              clearLineHandler interface
```

Comment:
But we can actually close the interface here, right?

Reply:
When I did, one of the processes eventually freezes from trying to write to stdout from what I assume some buffer filling and then the process waiting for it to be flushed. I think it was ogmios. Contracts would eventually freeze and it was very confusing why.

I switched to readline because it was concerning we were concatenating decoded partial unicode strings: https://github.com/purescript-node/purescript-node-streams/issues/37. I also considered running the filters on just a line rather than total output, which is still a possibility